### PR TITLE
Rawpart

### PR DESCRIPTION
**Pull request by hyc**
Add support for storing database on a raw device/partition. Can improve write performance.
Also may be useful for storing DB on removable storage (USB flash drives, etc.) for transporting
between online and offline daemons, and precludes the possibility of viruses hiding in a filesystem
(since there is no filesystem to hide in).

For use with monerod, would need to manually symlink &lt;data-dir&gt;/lmdb to the target device name.

This is only supported for POSIX systems, not WIN32. While there may be a way to access raw
devices in Windows, that usually requires Administrator access.

Note: currently monerod can't use this feature. Its startup checks for whether the target path is
a directory and if it supports mmap are interfering, they need to be skipped.

### CODE REVIEW

**Comment by hyc** [comment]
These checks must be disabled in order to use the PR:
[dif.txt](https://github.com/monero-project/monero/files/6777173/dif.txt)

**Comment by Malinero** [comment]
> For use with monerod, would need to manually symlink <data-dir>/lmdb to the target device name

What do you think about an option to avoid this manual operation ? stg like `--lmdb-device /dev/.....`

**Comment by hyc** [comment]
I think anybody who's smart enough to use this without accidentally destroying their live filesystem partitions is smart enough to set a symlink. And I feel there's less chance of accidentally typo'ing the wrong device name on the commandline, by relying on a manually created symlink that you can doublecheck a few times before starting up...

File: external/db_drivers/liblmdb/mdb.c [hunk 1]
```diff
@@ -4054,6 +4058,31 @@ mdb_env_read_header(MDB_env *env, int prev, MDB_meta *meta)
 	int			i, rc, off;
 	enum { Size = sizeof(pbuf) };
 
+	if (env->me_flags & MDB_RAWPART) {
+#define VM_ALIGN	0x200000
```
[the remark this answers is not in the document]
Comment (hyc) [reply external/db_drivers/liblmdb/mdb.c:4062]: The 2MB size is an optimal alignment for Flash memory erase blocks. Smaller may not be properly aligned for all devices, larger is not necessary.

**Comment by hyc** [comment]
Rebased onto current master

File: external/db_drivers/liblmdb/mdb.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (vtnerd) [reply external/db_drivers/liblmdb/mdb.c:4062]: Ahh ok I did in fact misunderstand this slightly, thanks for the clarification.

File: external/db_drivers/liblmdb/mdb.c [hunk 2]
```diff
@@ -4148,6 +4179,18 @@ mdb_env_init_meta(MDB_env *env, MDB_meta *meta)
 
 	psize = env->me_psize;
 
+	if ((env->me_flags & (MDB_RAWPART|MDB_WRITEMAP)) == (MDB_RAWPART|MDB_WRITEMAP)) {
```
[the remark this answers is not in the document]
Comment (vtnerd) [reply external/db_drivers/liblmdb/mdb.c:4182]: Yeah sorry, brainfart. Everything in *Nix is a file, so this is just opening the rawpart like any other file  (albeit one with a fixed size and location on disk) that gets written to like normal.

**Comment by vtnerd** [comment]
The only holdup is the lack of warning if mmap fails ... can/should that be left in? It's a somewhat useful warning to users, but maybe its ignored most of the time?

**Comment by hyc** [comment]
If mmap fails, LMDB will fail, and the daemon is dead in the water. It's not ignorable.

**Review by vtnerd — APPROVED** [review]

**Pull request merged**